src/common/runas.c

```diff
@@ -433,99 +739,238 @@ int run_as_noworker(enum run_as_cmd cmd,
 		goto end;
 	}
 	old_mask = umask(0);
-	ret = fct(data);
-	saved_errno = errno;
+	ret = fct(data, ret_value);
+	saved_errno = ret_value->_errno;
 	umask(old_mask);
 	errno = saved_errno;
 end:
 	return ret;
 }
 
 static
-int run_as(enum run_as_cmd cmd, struct run_as_data *data, uid_t uid, gid_t gid)
+int run_as_restart_worker(struct run_as_worker *worker)
 {
-	int ret;
+	int ret = 0;
+	char *procname = NULL;
+
+	procname = worker->procname;
+
+	/* Close socket to run_as worker process and clean up the zombie process */
+	run_as_destroy_worker();
+
+	/* Create a new run_as worker process*/
+	ret = run_as_create_worker(procname);
+	if (ret < 0 ) {
+		ERR("Restarting the worker process failed");
+		ret = -1;
+		goto err;
+	}
+err:
+	return ret;
+}
+
+static
+int run_as(enum run_as_cmd cmd, struct run_as_data *data,
+		   struct run_as_ret *ret_value, uid_t uid, gid_t gid)
+{
+	int ret, retry, saved_errno;
 
 	if (use_clone()) {
 		DBG("Using run_as worker");
-		pthread_mutex_lock(&worker_lock);
-		assert(global_worker);
-		ret = run_as_cmd(global_worker, cmd, data, uid, gid);
-		pthread_mutex_unlock(&worker_lock);
-
+		do {
+			retry = 0;
+			pthread_mutex_lock(&worker_lock);
+			assert(global_worker);
+
+			ret = run_as_cmd(global_worker, cmd, data, ret_value, uid, gid);
+			saved_errno = ret_value->_errno;
+
+			pthread_mutex_unlock(&worker_lock);
+			/*
+			 * If the worker thread crashed the errno is set to EIO. So we start
+			 * a new worker process and retry the command.
+			 */
+			if (ret == -1 && saved_errno == EIO) {
+				DBG("Socket closed unexpectedly... "
+					"Restarting the worker process");
+				ret = run_as_restart_worker(global_worker);
+
+				if (ret == -1) {
+					ERR("Failed to restart worker process.");
+					goto err;
+				}
+
+				retry = 1;
+			}
+		} while (retry);
```

Comment:
I would not retry the same command here. Report an error for the command and relaunch the run-as process so that the next commands can work.